### DISCUSSION TITLE
Build target curl 出错误

### DISCUSSION THREAD

**Discussion by keainye** [xmake-io/xmake#3620] [Q&A] [answered]
## 所用环境

msys2, mingw, qt, xmake

## 出错时执行的命令

`xmake config --yes --verbose --diagnosis --plat=mingw --mingw=/c/ProgramData/chocolatey/lib/mingw/tools/install/mingw64/ --qt=/c/Qt/6.5.0/mingw_64/`

## 问题描述

```shell
error: ...ake\repositories\xmake-repo\packages\l\libcurl\xmake.lua:140: ...gramdir\core\sandbox\modules\import\core\tool\linker.lua:75: @programdir\core\sandbox\modules\os.lua:378: execv(C:\ProgramData\chocolatey\lib\mingw\tools\install\mingw64\bin\x86_64-w64-mingw32-g++ -o C:\msys64\tmp\.xmake\230412\_B29EBCE7B9714B108ABD35D8E4CD0120.b C:\msys64\tmp\.xmake\230412\_B29EBCE7B9714B108ABD35D8E4CD0120.o -m64 -LC:\Users\shift\AppData\Local\.xmake\packages\l\libcurl\7.84.0\a7dd2c7b681d417487acfdea94efe991\lib -lcurl -ladvapi32 -lcrypt32 -lwldap32 -lwinmm -lws2_32 -luser32) failed(1)
stack traceback:
    [C]: in function 'error'
    [@programdir\core\base\os.lua:913]:
    [@programdir\core\sandbox\modules\os.lua:378]: in function 'execv'
    [@programdir\modules\core\tools\gcc.lua:408]:
    [C]: in function 'xpcall'
    [@programdir\core\base\utils.lua:280]:
    [@programdir\core\tool\linker.lua:222]: in function 'link'
    [...gramdir\core\sandbox\modules\import\core\tool\linker.lua:73]: in function 'link'
    [@programdir\modules\lib\detect\check_cxsnippets.lua:251]:
    [C]: in function 'xpcall'
    [@programdir\core\base\utils.lua:280]: in function 'trycall'
    [@programdir\core\sandbox\modules\try.lua:117]: in function 'try'
    [@programdir\modules\lib\detect\check_cxsnippets.lua:240]:
    [...ake\repositories\xmake-repo\packages\l\libcurl\xmake.lua:140]: in function 'script'
    [...dir\modules\private\action\require\impl\utils\filter.lua:125]: in function 'call'
    [...dir\modules\private\action\require\impl\actions\test.lua:41]:
    [...\modules\private\action\require\impl\actions\install.lua:366]:
    [C]: in function 'xpcall'
    [@programdir\core\base\utils.lua:280]: in function 'trycall'
    [@programdir\core\sandbox\modules\try.lua:117]: in function 'try'
    [...\modules\private\action\require\impl\actions\install.lua:287]:
    [...modules\private\action\require\impl\install_packages.lua:479]: in function 'jobfunc'
    [@programdir\modules\private\async\runjobs.lua:232]:
    [C]: in function 'xpcall'
    [@programdir\core\base\utils.lua:280]: in function 'trycall'
    [@programdir\core\sandbox\modules\try.lua:117]: in function 'try'
    [@programdir\modules\private\async\runjobs.lua:218]: in function 'cotask'
    [@programdir\core\base\scheduler.lua:404]:

  => install libcurl 7.84.0 .. failed
error: @programdir\modules\private\async\runjobs.lua:256: ...\modules\private\action\require\impl\actions\install.lua:428: install failed!
stack traceback:
    [C]: in function 'error'
    [@programdir\core\base\os.lua:913]:
    [...\modules\private\action\require\impl\actions\install.lua:428]: in function 'catch'
    [@programdir\core\sandbox\modules\try.lua:123]: in function 'try'
    [...\modules\private\action\require\impl\actions\install.lua:287]:
    [...modules\private\action\require\impl\install_packages.lua:479]: in function 'jobfunc'
    [@programdir\modules\private\async\runjobs.lua:232]:

stack traceback:
        [C]: in function 'error'
        @programdir\core\base\os.lua:913: in function 'os.raiselevel'
        (...tail calls...)
        @programdir\modules\private\async\runjobs.lua:256: in field 'catch'
        @programdir\core\sandbox\modules\try.lua:123: in global 'try'
        @programdir\modules\private\async\runjobs.lua:218: in upvalue 'cotask'
        @programdir\core\base\scheduler.lua:404: in function <@programdir\core\base\scheduler.lua:397>
```

## 尝试的解决方案

在 stack traceback 中我发现有 `os.raiselevel` 错误，故尝试以管理员身份启动 msys2，未解决。

在 [这个 xmake issue](https://github.com/xmake-io/xmake/issues/2395) 发现了相同的关键词 `os.raiselevel` 和 `cotask` ，按照其中的解决方案，将 xmake 升级至 dev，未解决。

求大佬帮忙看看 orz

**Comment by waruqi** [comment]
加 -vD 给下完整 logs

**Comment by keainye** [comment]
加了 -vD 之后 log 似乎没有发生变化。这是完整的 log，多谢大佬。

[xmake-log.txt](https://github.com/xmake-io/xmake/files/11208146/xmake-log.txt)

**Comment by waruqi** [comment, marked as answer]
> C:/ProgramData/chocolatey/lib/mingw/tools/install/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\shift\AppData\Local\.xmake\packages\l\libcurl\7.84.0\a7dd2c7b681d417487acfdea94efe991\lib/libcurl.a(version.c.obj):version.c:(.text+0x51): undefined reference to `psl_get_version'
C:/ProgramData/chocolatey/lib/mingw/tools/install/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\shift\AppData\Local\.xmake\packages\l\libcurl\7.84.0\a7dd2c7b681d417487acfdea94efe991\lib/libcurl.a(psl.c.obj):psl.c:(.text+0x31): undefined reference to `psl_free'
C:/ProgramData/chocolatey/lib/mingw/tools/install/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\shift\AppData\Local\.xmake\packages\l\libcurl\7.84.0\a7dd2c7b681d417487acfdea94efe991\lib/libcurl.a(psl.c.obj):psl.c:(.text+0xcd): undefined reference to `psl_latest'
C:/ProgramData/chocolatey/lib/mingw/tools/install/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\shift\AppData\Local\.xmake\packages\l\libcurl\7.84.0\a7dd2c7b681d417487acfdea94efe991\lib/libcurl.a(psl.c.obj):psl.c:(.text+0x107): undefined reference to `psl_free'
C:/ProgramData/chocolatey/lib/mingw/tools/install/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\shift\AppData\Local\.xmake\packages\l\libcurl\7.84.0\a7dd2c7b681d417487acfdea94efe991\lib/libcurl.a(psl.c.obj):psl.c:(.text+0x177): undefined reference to `psl_builtin'
C:/ProgramData/chocolatey/lib/mingw/tools/install/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\shift\AppData\Local\.xmake\packages\l\libcurl\7.84.0\a7dd2c7b681d417487acfdea94efe991\lib/libcurl.a(cookie.c.obj):cookie.c:(.text+0x13ff): undefined reference to `psl_is_cookie_domain_acceptable'

缺 psl 符号，curl 默认开了 psl 依赖编译。。可以提 pr 改进下 libcurl 包，加上 libpsl 依赖就行了

**Reply by keainye** [reply]
谢谢，我刚刚打开另一个 log 发现 `undefined reference to `psl_is_cookie_domain_acceptable'` ，应该是libpsl的问题了，我看看如何解决，感谢大佬

**Comment by waruqi** [comment]
我修复了，更新下仓库重装下

```
xrepo update-repo
xrepo remove --all libcurl
xmake f -c -p mingw
xmake
```

**Reply by keainye** [reply]
感谢，昨天把 msys2 环境中的 psl 删掉再装就解决了😂
